gnosis-imap/gnosis-imap-spec.ini

```diff
@@ -148,6 +148,89 @@ proxy_storage:
+    andBool OLD_OWNER =/=Int 0
+    andBool OLD_OWNER =/=Int SENTINEL
+    andBool INIT_PREV_OWNER ==Int OLD_OWNER
+    ; path condition
+    ; All owners are different
+    andBool SENTINEL   =/=Int PREV_OWNER // TODO: consider the case of SENTINEL ==Int PREV_OWNER  (by simply commenting out this line)
+  ; andBool SENTINEL   =/=Int OLD_OWNER  // required
+  ; andBool SENTINEL   =/=Int NEW_OWNER  // required
+    andBool PREV_OWNER =/=Int OLD_OWNER  // implied from `owners[prevOwner] == oldOwner` (implying `prevOwner != oldOwner` or `prevOwner == oldOwner == SENTINEL_OWNERS`) and `oldOwner != SENTINEL_OWNERS`
+    andBool PREV_OWNER =/=Int NEW_OWNER  // implied from `owners[prevOwner] == oldOwner != address(0)` and `owners[newOwner] == address(0)`
+    andBool OLD_OWNER  =/=Int NEW_OWNER  // implied from `owners[prevOwner] == oldOwner` (implying `owners[oldOwner] != address(0)`) and `owners[newOwner] == address(0)`
+    ; no hash collision between different owners
+    andBool #hashedLocation({COMPILER}, {OWNERS}, SENTINEL)   =/=Int #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, SENTINEL)   =/=Int #hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, SENTINEL)   =/=Int #hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER) =/=Int #hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER) =/=Int #hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER)  =/=Int #hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER)
+    ; no hash collision between owners and ownerCount
+    andBool #hashedLocation({COMPILER}, {OWNERS}, SENTINEL)   =/=Int #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER) =/=Int #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER)  =/=Int #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)
+    andBool #hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER)  =/=Int #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)
+ensures:
+;   ensures true
+;   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, SENTINEL))     ==Int INIT_SENTINEL
+;   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))   ==Int NEW_OWNER
+;   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER))    ==Int 0
+;   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER))    ==Int INIT_OLD_OWNER
+;   andBool select(M2, #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)) ==Int OWNERCOUNT
+;   andBool M1 ==IMap M2 except (
+;        SetItem(#hashedLocation({COMPILER}, {OWNERS}, NEW_OWNER))
+;        SetItem(#hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))
+;        SetItem(#hashedLocation({COMPILER}, {OWNERS}, OLD_OWNER))
+;       .Set)
```

Comment:
@anvacaru Your task is to prove this spec after uncommenting these ensure clauses.